cpp/ql/src/Likely Bugs/Memory Management/MemsetMayBeDeleted.ql

```diff
@@ -17,12 +13,24 @@
 
 import semmle.code.cpp.dataflow.TaintTracking
 
-from FunctionCall memset
+private predicate memsetName(string fn) {
+  fn = "memset" or
+  fn = "bzero" or
+  fn = "ZeroMemory" or
+  fn = "FillMemory" or
+  fn = "__builtin_memset"
+}
+
+from FunctionCall fc, Expr arg
 where
-  memset.getTarget().getName() = "memset" or memset.getTarget().getName() = "ZeroMemory" and
-  not exists(TaintTracking::Configuration cf, DataFlow::Node source, DataFlow::Node sink |
-    cf.hasFlow(source, sink) and
-    source.asExpr() = memset.getArgument(1)
+  memsetName(fc.getTarget().getName()) and
+  arg = fc.getArgument(0) and
+  not exists(Expr succ |
+    TaintTracking::localTaint(DataFlow::definitionByReferenceNodeFromArgument(arg),
+      DataFlow::exprNode(succ))
+  ) and
+  not exists(Parameter parm |
+    TaintTracking::localTaint(DataFlow::parameterNode(parm),
+      DataFlow::exprNode(arg))
```

Comment:
This checking of flow from `Parameter` is just one case out of the many ways that `arg` can be aliased. It could be a whack-a-mole game to enumerate them all. I suggest that we make the query the other way around: give an alert only if there's provably no way that the memory could be read after being cleared. That means, to begin with, that we only support stack-allocated arrays.

The easy way to check whether a variable may escape is to use https://github.com/Semmle/ql/blob/master/cpp/ql/src/semmle/code/cpp/dataflow/EscapesTree.qll#L254. Unfortunately it's very conservative, so it would be better to use the IR.

Reply:
> I suggest that we make the query the other way around: give an alert only if there's provably no way that the memory could be read after being cleared. 

How would one obtain such a proof?  I'm in the dark here.

Reply:
> The easy way to check whether a variable may escape is to use https://github.com/Semmle/ql/blob/master/cpp/ql/src/semmle/code/cpp/dataflow/EscapesTree.qll#L254. Unfortunately it's very conservative, so it would be better to use the IR.

I'm still wrapping my head around this one.  So does the `VariableAccess` parameter correspond to the pointer/array variable that is the first argument to `memset`? What about the `Expr` parameter?  The description in the sources is underwhelming.  I'm looking at `ReturnStackAllocatedMemory.ql` and it seems that the second parameter _could_ correspond to the value of the `return` statement, but not always.

Reply:
> Unfortunately it's very conservative, so it would be better to use the IR.

So perhaps it's time for me to start learning about the IR.